Method References - the unpolished gem of Java 8

Jezorko opened the pull request.

Commit:
Add blog post "Method References - the unpolished gem of Java 8"

gastonfournier commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,299 @@
+---
+layout: post
+title:  "Method References - the unpolished gem of Java 8"
```
Checking if I can comment

Commit:
Apply Gaston's comments

Context: https://github.com/Jezorko/Jezorko.github.io/pull/2

larribas commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,299 @@
+
+## Why Method References are great
+
+For many, method references are just a bit of synthetic sugar sprinkled about in the language.
```
synthetic => syntactic sugar ? or is it intentional?

Jezorko replied:
Not intentional. Good catch, thanks! (:
Fixed here: https://github.com/Jezorko/Jezorko.github.io/pull/1/commits/478efeefcbe646c917d5163395c5c22c1e52861f

larribas commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,299 @@
+    .forEach(veterinarian::treat);
+{% endhighlight %}
+
+The two examples above are obviously not flawless.
```
I actually find them worse than the lambda hehe

Jezorko replied:
It's a matter of opinion then, I guess (;
Still better to have more than one option available!

larribas reviewed (commented):
Hey, I enjoyed reading the article!

Let's speak about it in more detail later ;)

Commit:
Fix typo synthetic => syntactic

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r349554966

Commit:
Replace "easy to implement in JDK" statement with an example from Kotlin

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+
+If you were to pick your favorite year from the entire history of mankind, which year would it be?
+Most people would probably say 1932, when [the Great Emu War](https://www.nationalgeographic.com.au/history/the-great-australian-emu-war.aspx) was fought in Australia.
+This one is definitely in my top 10; however it falls behind 2014 - the year when Java 8 has been released.
```
```suggestion
This one is definitely in my top 10; however it falls behind 2014 - the year when Java 8 was released.
```

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+Most people would probably say 1932, when [the Great Emu War](https://www.nationalgeographic.com.au/history/the-great-australian-emu-war.aspx) was fought in Australia.
+This one is definitely in my top 10; however it falls behind 2014 - the year when Java 8 has been released.
+
+Among other things, this glorious language update added some functional programming capabilities to the language.
```
```suggestion
Among other things, this glorious update added some functional programming capabilities to the language.
```
Don't Repeat Yourself :wink:

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+
+[Method References](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.13)—also introduced in Java 8—provide a different syntax for Lambda expressions.
+They may be used with any type of method: instance, static or a constructor.
+Although often overlooked, I believe they have a power to greatly improve our code.
```
```suggestion
Although often overlooked, I believe they have the power to greatly improve our code.
```

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+They may be used with any type of method: instance, static or a constructor.
+Although often overlooked, I believe they have a power to greatly improve our code.
+
+I will not talk about how to use Method References here.
```
```suggestion
I will not talk about how to use Method References here, since this has.... etc.
```
flows a bit easier

Jezorko replied:
Fixed here: https://github.com/Jezorko/Jezorko.github.io/pull/1/commits/02dcaada2c230dec5dac15a16f1f93e6c45608b3 (:

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+
+For many, method references are just a bit of syntactic sugar sprinkled about in the language.
+Sometimes they will accept their IDE's suggestion and transform a Lambda into one.
+A nice to have addition, at most.
```
```suggestion
A nice-to-have addition, at most.
```

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+I counted four: `dog`, `optionalDogHouse`, `dogHouse` and `treats`.
+
+One of these names bothers me in particular.
+Is the `dogHouse` still `optional` after we already checked its' presence?
```
```suggestion
Is the `dogHouse` still `optional` after we already checked its presence?
```

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+
+## How Method References could be even better
+
+As useful of a feature as Method References are, it surprises me that they are so limited.
```
```suggestion
As useful as Method References are, it surprises me that they are so limited.
```
it's ok to be concise here, I think

Jezorko replied:
Yep, that reads better :+1:

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+### Calling functional interface's methods from the method reference
+
+Sometimes the object's method alone are not enough to implement the logic.
+In these cases, we can take one of the few approaches:
```
```suggestion
In these cases, we can take one of a few approaches:
```

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+
+ 1. create a decorator around the object and add missing methods
+ 2. write a new method that takes the object as an argument
+ 3. put the code in-line
```
```suggestion
 3. in-line the code
```

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+ 4. extend the offending class
+
+I listed them in order of my own preference.
+Each of these has their own advantages and disadvantages.
```
```suggestion
I listed them in order of my own preference, but each of these has their own advantages and disadvantages.
```

Jezorko committed:
Fix "has been" -> "was"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351218388

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+The second one forces us to ask: where do we stop?
+ * Should we allow multiple static methods to be chained, or just one?
+ * If they can be chained, then the feature could be potentially misused to produce unreadable code.
+ * If only a one can be used, then we are ridding ourselves of one limitation only to add another.
```
```suggestion
 * If only one can be used, then we are ridding ourselves of one limitation only to add another.
```

Jezorko committed:
Fix repetition

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351218648

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Fix "a" -> "the"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351219093

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+
+A similar feature is already available in Kotlin, called [single-expression functions](https://kotlinlang.org/docs/reference/functions.html#single-expression-functions).
+Although they cannot accept method references, these functions can significantly reduce the amount of boilerplate.
+This is how our decorator class could look like if it were written in Kotlin: 
```
yessssss I was hoping you'd mention Kotlin!

huwdaviesglovo commented on _posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown:
```diff
@@ -0,0 +1,311 @@
+Your code will become cleaner and easier to refactor, at no extra cost.
+Keep in mind though that they have some limitations and unless those get waived in the next Java releases, some workarounds need to be implemented.
+
+Method references are—in my humble opinion—not a complete feature.
```
```suggestion
Method references are - in my humble opinion - not a complete feature.
```

Jezorko replied:
These are em dashes though!

huwdaviesglovo replied:
interesting... I didn't know about them. Maybe that's a sign they're not universal enough? 😉

huwdaviesglovo commented:
nice article! Let me know if you have any questions about my suggestions 😄

bougueddach commented:
❤️

Jezorko committed:
Add dashes in "nice-to-have"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351219845

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Remove ' in "its'"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351220390

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Remove "of a feature"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351220883

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Replace "the" with "a"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351221475

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Fix "in-line" usage

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351222327

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Fix wording

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351222628

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Jezorko committed:
Remove unnecessary "a"

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351223638

Co-Authored-By: Huw Davies <53854805+huwdaviesglovo@users.noreply.github.com>

Commit:
Flow fix

Context: https://github.com/Jezorko/Jezorko.github.io/pull/1#discussion_r351219415

Commit:
Move post to actual release date
